File: src/main/java/net/moltendorf/Bukkit/IntelliDoors/controller/TrapDoor.java
package net.moltendorf.Bukkit.IntelliDoors.controller;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Sound;
import org.bukkit.block.Block;

/**
 * Created by moltendorf on 15/05/23.
 */
public class TrapDoor implements Door {
  public static Door getDoor(Block block) {
    return null;
  }

  private Material type;

  private Location location;

  protected Block block;

  protected byte data;

  public TrapDoor(Block block) {
    this.block = block;

    data = block.getData();
    location = block.getLocation();
    type = block.getType();
  }

  @Override
  public void open() {
    if (isClosed()) {
      data += 4;

      block.setData(data);
    }
  }

  @Override
  public void close() {
    if (isOpened()) {
      data -= 4;

      block.setData(data);
    }
  }

  @Override
  public void toggle() {
    data += isClosed() ? 4 : -4;

    block.setData(data);
  }

  @Override
  public boolean isOpened() {
    return (data & 4) == 4;
  }

  @Override
  public boolean isClosed() {
    return (data & 4) == 0;
  }

  @Override
  public Material getType() {
    return type;
  }

  @Override
  public void wasToggled(Door onDoor) {
    if (type == Material.IRON_TRAPDOOR) {
      if (isClosed()) {
        open();

        location.getWorld().playSound(location, Sound.BLOCK_IRON_TRAPDOOR_OPEN, 1, 1);
      } else {
        close();

        location.getWorld().playSound(location, Sound.BLOCK_IRON_TRAPDOOR_CLOSE, 1, 1);
      }
    }
  }
}

File: src/main/kotlin/net/moltendorf/Bukkit/IntelliDoors/controller/TrapDoor.kt
package net.moltendorf.Bukkit.IntelliDoors.controller

import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.block.Block

/**
 * Created by moltendorf on 15/05/23.
 */
class TrapDoor(block: Block) : AbstractDoor(block) {
  override fun sound(open: Boolean): Sound {
    return when (type) {
      Material.IRON_TRAPDOOR -> {
        if (open) Sound.BLOCK_IRON_TRAPDOOR_OPEN else Sound.BLOCK_IRON_TRAPDOOR_CLOSE
      }
      Material.TRAP_DOOR -> {
        if (open) Sound.BLOCK_WOODEN_TRAPDOOR_OPEN else Sound.BLOCK_WOODEN_TRAPDOOR_CLOSE
      }
      else -> Sound.BLOCK_ANVIL_LAND
    }
  }

  companion object {
    operator fun get(block: Block): TrapDoor? {
      return when (block.type) {
        Material.IRON_TRAPDOOR,
        Material.TRAP_DOOR -> TrapDoor(block)
        else -> null
      }
    }
  }
}
